### PR TITLE
[FLOC-3108] A big composeable AMP argument type

### PR DESCRIPTION
**Pull request by wallrj**
Fixes: https://clusterhq.atlassian.net/browse/FLOC-3108

### CODE REVIEW

File: flocker/control/_protocol.py [hunk 1]
```diff
@@ -55,6 +58,62 @@
...
+        method of the wrapped ``Argument`` is then called supplied with the
+        updated ``strings`` dictionary, deserializes the large value and
+        populates the ``objects`` dictionary with the result.
+
+        See ``IArgumentType`` for argument and return type documentation.
+        """
+
+        value = BytesIO()
+        for counter in count(0):
+            chunk = strings.get("%s.%d" % (name, counter))
```
Comment (wallrj) [inline flocker/control/_protocol.py]: We could `pop` here instead to avoid having two copies of the bytes in memory.

Comment (itamarst) [inline flocker/control/_protocol.py]: Mutation is a bad idea; we should avoid if possible.

**Comment by itamarst** [comment]
I merged forward to deal with some conflicts.

File: flocker/control/_protocol.py (hunk 1 above)
Comment (itamarst) [inline flocker/control/_protocol.py]: Moreover this doesn't copy, it just adds another reference to the same string.

Comment (itamarst) [inline flocker/control/_protocol.py]: Python never copies stuff unless you explicitly say so with `.copy()` or `[:]` or `copy.copy` or whatever.

Comment (itamarst) [inline flocker/control/_protocol.py]: Oh, you mean a copy in `BytesIO`. And `strings` is being severely mutated anyway. Probably fine as is.

**Comment by itamarst** [comment]
OK, merge when green.

**Pull request merged**